File: week-1-c/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool check(long number); // abstracted the Luhn's algorithm part of verification
int main(void)

{

    // declaring variables
    long card_number;
    int length = 0, digit = 0, double_digits = 0;
    bool verification;

    // prompting the user for the card number
    card_number = get_long("Number: ");

    // verifying number validity
    verification = check(card_number);
    if (verification == false)
    {

        // this ends the program if the Luhn's Algorithm verification is false
        return 0;
    }
    while (card_number > 0)
    {
        if (card_number < 10)
        {
            digit = card_number;
        }
        if (card_number > 10 && card_number < 100)
        {
            double_digits = card_number;
        }

        // takes digits away one at a time to operate both the counter, and the first digit identifier
        card_number /= 10;

        // counts how long the card number is
        length++;
    }

    // final verification process, determining what brand  of card it is
    if (length == 15 && (double_digits == 34 || double_digits == 37))
    {
        printf("AMEX\n");
        return 0;
    }
    else if (length == 16 && (double_digits == 51 ||
                              double_digits == 52 ||
                              double_digits == 53 ||
                              double_digits == 54 ||
                              double_digits == 55))
    {
        printf("MASTERCARD\n");
        return 0;
    }
    else if ((length == 13 || length == 16) && digit == 4)
    {
        printf("VISA\n");
        return 0;
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

bool check(long number)
{

    // sum is the total number which has to end in a 0. holder is a placeholder that gets added to the sum
    int sum = 0, holder = 0;
    bool x = true;
    while (number > 0)
    {

        // the card number always starts as true
        if (x == true)
        {

            // the last digit is separated and added to the total
            sum += number % 10;
            number /= 10;

            // this starts the next loop so every other number can be multiplied by 2
            x = false;
        }
        else
        {
            holder = number % 10;

            // multiplying the previously second-to-last number by 2
            holder *= 2;
            if (holder >= 10)
            {

                // this is so we can separate and add the individual digits in a number over 10
                sum += holder % 10;
                sum += holder / 10;
            }
            else
            {

                // this is where the sum and the holder are added together
                sum += holder;
            }

            // resetting the loop until all numbers have been separated and added together accordingly
            x = true;
            number /= 10;
        }
    }

    // the card is valid if the remainder of the sum / 10 is 0!
    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        printf("INVALID\n");
        return false;
    }
}

File: week-6-python/sentimental-credit/credit.py
from cs50 import get_int
import sys
from math import floor


def main():

    # declare variables and get user input
    length = 0
    digit = 0
    double_digits = 0
    verification = True

    card_number = get_int("Number: ")

    # check if the card number is valid and exit if not
    verification = check(card_number)

    if verification == False:
        sys.exit(1)

    # keep track of card length and first 2 digits
    while card_number > 0:

        if card_number < 10:
            digit = card_number

        if card_number > 10 and card_number < 100:
            double_digits = card_number

        card_number = floor(card_number / 10)
        length += 1

    # check for card type based on length and digits
    if length == 15 and (double_digits == 34 or double_digits == 37):
        print(f"AMEX\n")
        sys.exit(0)

    elif length == 16 and (double_digits == 51 or
                           double_digits == 52 or
                           double_digits == 53 or
                           double_digits == 54 or
                           double_digits == 55):
        print(f"MASTERCARD\n")
        sys.exit(0)

    elif (length == 13 or length == 16) and digit == 4:
        print(f"VISA\n")
        sys.exit(0)

    else:
        print("INVALID")
        sys.exit(2)


def check(number):

    # declare variables
    sum = 0
    holder = 0
    x = True

    # iterate through the entire card number sequence
    while number > 0:

        # add every other last digit to the sum
        if x == True:

            sum += number % 10
            number = floor(number / 10)
            x = False

        else:

            # multiply every other last digit by 2 and add those digits to the sum
            holder = number % 10
            holder *= 2

            # adds individual digits to the sum if the product of holder * 2 is > 10
            if holder >= 10:

                sum += holder % 10
                sum += floor(holder / 10)

            else:

                sum += holder

            # resets x to True and resets the loop
            x = True
            number = floor(number / 10)

    # check if the sum is divisible by 10 and return false if not
    if sum % 10 == 0:
        return True

    else:
        print("INVALID")
        return False


if __name__ == "__main__":
    main()